File: framework/languages/dotnet/e2e/ResilienceLifecycle/Client/Scenarios/RlD2ObserverFaultScenario.cs
using ResilienceLifecycle.Client.Support;
using ResilienceLifecycle.Shared;
using Zlink.HttpClient;

namespace ResilienceLifecycle.Client.Scenarios;

// RL-D2 verifies dispatch observer fault isolation and continued messaging.
internal static class RlD2ObserverFaultScenario
{
    public static async Task RunAsync(
        ZLinkHttpClient consumer,
        ZLinkHttpClient providerA,
        ZLinkHttpClient providerB)
    {
        await providerB.Post("/admin/fault/observer-throws").SubmitRawAsync();
        var missing = await consumer.Post("/profile/request/missing")
            .Body(new ProfileRequest("fast", "rl-d2-error"))
            .SubmitRawAsync();
        ScenarioAssert.That(missing.Status >= 500, "RL-D2 missing handler request should fail.");

        var followUp = (await consumer.Post("/profile/request")
            .Body(new ProfileRequest("fast", "rl-d2-after"))
            .SubmitAsync<ProfileReply>()).Body;
        ScenarioAssert.That(followUp.Value == "profile:fast",
            "RL-D2 messaging did not continue after observer failure.");
        await providerB.Post("/admin/fault/none").SubmitRawAsync();

        {
            using var timeout = new CancellationTokenSource(TimeSpan.FromSeconds(15));
            var waitA = providerA.Post("/evidence/wait").Body(new EvidenceWaitRequest(["marker=rl-d2-after"], []))
                .SubmitAsync<string[]>(timeout.Token).AsTask();
            var waitB = providerB.Post("/evidence/wait").Body(new EvidenceWaitRequest(["marker=rl-d2-after"], []))
                .SubmitAsync<string[]>(timeout.Token).AsTask();
            var completed = await Task.WhenAny(waitA, waitB);
            var evidence = (await completed).Body;
            timeout.Cancel();
            ScenarioAssert.That(evidence.Any(line => line.Contains("marker=rl-d2-after", StringComparison.Ordinal)),
                "RL-D2 did not record expected evidence 'marker=rl-d2-after'.");
        }

        Console.WriteLine("scenario RL-D2 passed");
    }
}

File: framework/languages/node/e2e/ResilienceLifecycle/Client/Scenarios/rl-d2-observer-fault-scenario.ts
// RL-D2: observer 실패 격리 시나리오를 검증한다.
import type { ProfileRes, RequestFailureRes } from '../../Shared/messages';
import type { ClientOptions } from '../Support/client-options';
import { getJson, postJson } from '../../../http-client';
import { profileReq } from '../Support/resilience-helpers';
import { ensure } from '../Support/scenario-assert';

export async function runRlD2(options: ClientOptions): Promise<void> {
  await postJson(options.providerAUrl, '/admin/fault/observer-throws');
  await postJson(options.providerBUrl, '/admin/fault/observer-throws');

  const missing = await postJson<RequestFailureRes>(
    options.consumerUrl,
    '/profile/missing-request',
    profileReq('rl-d2-error')
  );
  ensure(missing.failed, 'RL-D2 missing handler request should fail.');

  await waitForEitherEvidence(options, 'dispatch-error', 'RL-D2 did not record dispatch-error evidence.');

  const followUp = await postJson<ProfileRes>(
    options.consumerUrl,
    '/profile/request',
    profileReq('rl-d2-after')
  );
  ensure(followUp.value === 'profile:fast', 'RL-D2 messaging did not continue after observer failure.');

  await postJson(options.providerAUrl, '/admin/fault/none');
  await postJson(options.providerBUrl, '/admin/fault/none');

  await waitForEitherEvidence(options, 'marker=rl-d2-after', 'RL-D2 did not record expected follow-up evidence.');

  console.log('scenario RL-D2 passed');
}

async function waitForEitherEvidence(options: ClientOptions, contains: string, message: string): Promise<void> {
  const deadline = Date.now() + 15000;
  while (Date.now() < deadline) {
    const snapshots = await Promise.allSettled([
      getJson<string[]>(options.providerAUrl, '/evidence'),
      getJson<string[]>(options.providerBUrl, '/evidence')
    ]);
    if (snapshots.some((snapshot) =>
      snapshot.status === 'fulfilled' && snapshot.value.some((line) => line.includes(contains)))) {
      return;
    }
    await new Promise((resolve) => setTimeout(resolve, 250));
  }
  ensure(false, message);
}
